File: src/main/java/org/mitallast/queue/common/file/FileService.java
package org.mitallast.queue.common.file;

import com.google.common.base.Preconditions;
import com.google.inject.Inject;
import com.typesafe.config.Config;

import java.io.*;
import java.nio.file.*;
import java.util.Iterator;
import java.util.stream.Stream;

public class FileService {
    private final File root;

    @Inject
    public FileService(Config config) {
        File path = new File(config.getString("node.path"));
        root = new File(path, config.getString("transport.port")).getAbsoluteFile();
        if (!root.exists() && !root.mkdirs()) {
            throw new IOError(new IOException("error create directory: " + root));
        }
    }

    public File service(String service) {
        Path rootPath = root.toPath().normalize();
        Path servicePath = Paths.get(root.getPath(), service).normalize();
        Preconditions.checkArgument(servicePath.startsWith(rootPath), "service path");
        return servicePath.toFile();
    }

    public File resource(String service, String key) {
        Path servicePath = service(service).toPath();

        Path filePath = Paths.get(servicePath.toString(), key).normalize();
        Preconditions.checkArgument(filePath.startsWith(servicePath), "resource path");

        File resource = filePath.toFile();
        if (!resource.exists()) {
            if (!resource.getParentFile().exists()) {
                if (!resource.getParentFile().mkdirs()) {
                    throw new IOError(new IOException("Error create directory " + resource.getParentFile()));
                }
            }
            try {
                if (!resource.createNewFile()) {
                    throw new IOError(new IOException("Error create file " + resource));
                }
            } catch (IOException e) {
                throw new IOError(e);
            }
        }
        return resource;
    }

    public File temporary(String service, String prefix, String suffix) {
        Path servicePath = service(service).toPath();
        try {
            return Files.createTempFile(servicePath, prefix, suffix).toFile();
        } catch (IOException e) {
            throw new IOError(e);
        }
    }

    public Stream<Path> resources(String service) {
        Path servicePath = service(service).toPath();
        if (!servicePath.toFile().exists()) {
            return Stream.empty();
        }
        try {
            return Files.walk(servicePath)
                .filter(path -> path.toFile().isFile())
                .map(servicePath::relativize);
        } catch (IOException e) {
            throw new IOError(e);
        }
    }

    public Stream<Path> resources(String service, String prefix) {
        PathMatcher matcher = FileSystems.getDefault().getPathMatcher(prefix);

        Path servicePath = service(service).toPath();
        if (!servicePath.toFile().exists()) {
            return Stream.empty();
        }
        try {
            return Files.walk(servicePath)
                .filter(path -> path.toFile().isFile())
                .map(servicePath::relativize)
                .filter(matcher::matches);
        } catch (IOException e) {
            throw new IOError(e);
        }
    }

    public void delete(String service) {
        Path servicePath = service(service).toPath();
        if (!servicePath.toFile().exists()) {
            return;
        }
        try {
            Iterator<File> iterator = Files.walk(servicePath)
                .map(Path::toFile)
                .sorted((o1, o2) -> -o1.compareTo(o2))
                .iterator();
            while (iterator.hasNext()) {
                File next = iterator.next();
                delete(next);
            }
        } catch (IOException e) {
            e.printStackTrace();
        }

    }

    public void delete(File file) {
        if (file.exists() && !file.delete()) {
            throw new IOError(new IOException("Error delete file " + file));
        }
    }

    public void move(File tmp, File dest) {
        try {
            Files.move(tmp.toPath(), dest.toPath(), StandardCopyOption.REPLACE_EXISTING);
        } catch (IOException e) {
            throw new IOError(e);
        }
    }

    public DataOutputStream output(File file) {
        return output(file, false);
    }

    public DataOutputStream output(File file, boolean append) {
        try {
            FileOutputStream outputStream = new FileOutputStream(file, append);
            BufferedOutputStream bufferedOutputStream = new BufferedOutputStream(outputStream);
            return new DataOutputStream(bufferedOutputStream);
        } catch (IOException e) {
            throw new IOError(e);
        }
    }

    public DataInputStream input(File file) {
        try {
            FileInputStream inputStream = new FileInputStream(file);
            BufferedInputStream bufferedInputStream = new BufferedInputStream(inputStream);
            return new DataInputStream(bufferedInputStream);
        } catch (IOException e) {
            throw new IOError(e);
        }
    }
}

File: src/main/java/org/mitallast/queue/common/file/FileService.kt
package org.mitallast.queue.common.file

import com.google.common.base.Preconditions
import com.google.inject.Inject
import com.typesafe.config.Config

import java.io.*
import java.nio.file.*
import java.util.stream.Stream

class FileService @Inject
constructor(config: Config) {
    private val root: File

    init {
        val path = File(config.getString("node.path"))
        root = File(path, config.getString("transport.port")).absoluteFile
        if (!root.exists() && !root.mkdirs()) {
            throw IOException("error create directory: " + root)
        }
    }

    fun service(service: String): File {
        val rootPath = root.toPath().normalize()
        val servicePath = Paths.get(root.path, service).normalize()
        Preconditions.checkArgument(servicePath.startsWith(rootPath), "service path")
        return servicePath.toFile()
    }

    fun resource(service: String, key: String): File {
        val servicePath = service(service).toPath()

        val filePath = Paths.get(servicePath.toString(), key).normalize()
        Preconditions.checkArgument(filePath.startsWith(servicePath), "resource path")

        val resource = filePath.toFile()
        if (!resource.exists()) {
            if (!resource.parentFile.exists()) {
                if (!resource.parentFile.mkdirs()) {
                    throw IOException("Error create directory " + resource.parentFile)
                }
            }
            if (!resource.createNewFile()) {
                throw IOException("Error create file " + resource)
            }
        }
        return resource
    }

    fun temporary(service: String, prefix: String, suffix: String): File {
        val servicePath = service(service).toPath()
        return Files.createTempFile(servicePath, prefix, suffix).toFile()
    }

    fun resources(service: String): Stream<Path> {
        val servicePath = service(service).toPath()
        if (!servicePath.toFile().exists()) {
            return Stream.empty()
        }
        return Files.walk(servicePath)
            .filter { path -> path.toFile().isFile }
            .map { servicePath.relativize(it) }
    }

    fun resources(service: String, prefix: String): Stream<Path> {
        val matcher = FileSystems.getDefault().getPathMatcher(prefix)

        val servicePath = service(service).toPath()
        if (!servicePath.toFile().exists()) {
            return Stream.empty()
        }
        return Files.walk(servicePath)
            .filter { path -> path.toFile().isFile }
            .map { servicePath.relativize(it) }
            .filter { matcher.matches(it) }
    }

    fun delete(service: String) {
        val servicePath = service(service).toPath()
        if (servicePath.toFile().exists()) {
            val iterator = Files.walk(servicePath)
                .map { it.toFile() }
                .sorted { o1, o2 -> -o1.compareTo(o2) }
                .iterator()
            while (iterator.hasNext()) {
                val next = iterator.next()
                delete(next)
            }
        }
    }

    fun delete(file: File) {
        if (file.exists() && !file.delete()) {
            throw IOException("Error delete file " + file)
        }
    }

    fun move(tmp: File, dest: File) {
        Files.move(tmp.toPath(), dest.toPath(), StandardCopyOption.REPLACE_EXISTING)
    }

    fun output(file: File, append: Boolean = false): DataOutputStream {
        try {
            val outputStream = FileOutputStream(file, append)
            val bufferedOutputStream = BufferedOutputStream(outputStream)
            return DataOutputStream(bufferedOutputStream)
        } catch (e: IOException) {
            throw IOError(e)
        }

    }

    fun input(file: File): DataInputStream {
        try {
            val inputStream = FileInputStream(file)
            val bufferedInputStream = BufferedInputStream(inputStream)
            return DataInputStream(bufferedInputStream)
        } catch (e: IOException) {
            throw IOError(e)
        }

    }
}
